File: py/agentctl/skill_install.py
"""Install the bundled agentctl skill without replacing unreviewed content."""
from __future__ import annotations

import json
import os
import pwd
import selectors
import signal
import stat
import subprocess
import tempfile
import time
from importlib.resources import files
from pathlib import Path
from typing import Sequence

from agentctl.errors import AgentDeliveryError

HARNESSES = ("codex", "claude", "muse")
_INSTALL_OUTPUT_BYTES = 1 << 20
_INSTALL_TIMEOUT = 30.0


def _root(harness: str) -> Path:
    override = os.environ.get(f"AGENTCTL_{harness.upper()}_SKILLS_DIR")
    if override is not None:
        root = Path(override)
        if not root.is_absolute():
            raise AgentDeliveryError(f"{harness} skill directory override must be absolute")
        return root
    home = Path(pwd.getpwuid(os.getuid()).pw_dir)
    return home / {"codex": ".codex", "claude": ".claude"}[harness] / "skills"


def _ensure_real_directory(path: Path) -> None:
    """Create missing components while refusing every symlink in the destination path."""
    if not path.is_absolute():
        raise AgentDeliveryError(f"skill destination must be absolute: {path}")
    current = Path(path.anchor)
    for part in path.parts[1:]:
        current /= part
        try:
            metadata = current.lstat()
        except FileNotFoundError:
            try:
                current.mkdir(mode=0o700)
            except OSError as exc:
                raise AgentDeliveryError(
                    f"cannot create skill destination {current}: {exc}"
                ) from exc
            continue
        except OSError as exc:
            raise AgentDeliveryError(f"cannot inspect skill destination {current}: {exc}") from exc
        if stat.S_ISLNK(metadata.st_mode) or not stat.S_ISDIR(metadata.st_mode):
            raise AgentDeliveryError(f"refusing non-directory skill destination: {current}")


def _existing(path: Path, content: str, *, force: bool) -> str | None:
    """Return unchanged, or validate that native/direct replacement is allowed."""
    if not os.path.lexists(path):
        return None
    metadata = path.lstat()
    if stat.S_ISLNK(metadata.st_mode) or not stat.S_ISREG(metadata.st_mode):
        raise AgentDeliveryError(f"refusing non-regular skill file: {path}")
    try:
        existing = path.read_text(encoding="utf-8")
    except (OSError, UnicodeError) as exc:
        raise AgentDeliveryError(f"cannot inspect installed skill {path}: {exc}") from exc
    if existing == content:
        return "unchanged"
    if not force:
        raise AgentDeliveryError(f"refusing to overwrite divergent skill {path}; inspect it or pass --force")
    return None


def _refuse_directory_symlink(path: Path) -> None:
    if not os.path.lexists(path):
        return
    metadata = path.lstat()
    if stat.S_ISLNK(metadata.st_mode) or not stat.S_ISDIR(metadata.st_mode):
        raise AgentDeliveryError(f"refusing non-directory skill destination: {path}")


def _write(path: Path, content: str, *, force: bool) -> str:
    directory = path.parent
    _ensure_real_directory(directory)
    outcome = _existing(path, content, force=force)
    if outcome is not None:
        return outcome
    fd, temporary = tempfile.mkstemp(prefix=".SKILL.md.", dir=directory, text=True)
    try:
        os.fchmod(fd, 0o600)
        with os.fdopen(fd, "w", encoding="utf-8") as stream:
            stream.write(content)
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(temporary, path)
        descriptor = os.open(directory, os.O_RDONLY | os.O_DIRECTORY)
        try:
            os.fsync(descriptor)
        finally:
            os.close(descriptor)
    finally:
        try:
            os.unlink(temporary)
        except FileNotFoundError:
            pass
    return "installed"


def _muse_config_home() -> Path:
    configured = os.environ.get("XDG_CONFIG_HOME")
    if configured is not None:
        root = Path(configured)
        if not root.is_absolute():
            raise AgentDeliveryError("XDG_CONFIG_HOME must be absolute for Muse skill installation")
        return root / "muse"
    return Path(pwd.getpwuid(os.getuid()).pw_dir) / ".config" / "muse"


def _muse_executable_candidates() -> tuple[Path, ...]:
    home = Path(pwd.getpwuid(os.getuid()).pw_dir)
    return (
        Path("/usr/local/bin/muse"),
        Path("/usr/bin/muse"),
        home / ".local/bin/muse",
        home / "bin/muse",
        home / ".cargo/bin/muse",
    )


def _muse_executable() -> str:
    configured = os.environ.get("AGENTCTL_MUSE_BIN")
    candidate = configured if configured is not None else next((
        str(path) for path in _muse_executable_candidates()
        if path.is_file() and os.access(path, os.X_OK)
    ), None)
    if not candidate:
        raise AgentDeliveryError("Muse executable not found; install Muse before installing its skill")
    path = Path(candidate)
    if not path.is_absolute():
        raise AgentDeliveryError("AGENTCTL_MUSE_BIN must be an absolute path")
    resolved = Path(os.path.realpath(path))
    try:
        metadata = resolved.stat()
    except OSError as exc:
        raise AgentDeliveryError(f"cannot inspect Muse executable {resolved}: {exc}") from exc
    if (not stat.S_ISREG(metadata.st_mode) or not os.access(resolved, os.X_OK)
            or metadata.st_mode & (stat.S_IWGRP | stat.S_IWOTH)):
        raise AgentDeliveryError(f"refusing unsafe Muse executable: {resolved}")
    return str(resolved)


def _run_muse_installer(command: list[str], environment: dict[str, str]) -> tuple[int, str, str]:
    """Run the native manager with bounded pipes, time, and process-group cleanup."""
    try:
        process = subprocess.Popen(
            command, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE,
            stderr=subprocess.PIPE, env=environment, start_new_session=True,
        )
    except OSError as exc:
        raise AgentDeliveryError(f"Muse skill installation failed: {exc}") from exc
    assert process.stdout is not None and process.stderr is not None
    streams = (process.stdout, process.stderr)
    buffers = {stream.fileno(): bytearray() for stream in streams}
    selector = selectors.DefaultSelector()
    deadline = time.monotonic() + _INSTALL_TIMEOUT
    try:
        for stream in streams:
            os.set_blocking(stream.fileno(), False)
            selector.register(stream.fileno(), selectors.EVENT_READ)
        while selector.get_map():
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise subprocess.TimeoutExpired(command, _INSTALL_TIMEOUT)
            for key, _events in selector.select(min(0.1, remaining)):
                try:
                    chunk = os.read(key.fd, 65_536)
                except BlockingIOError:
                    continue
                if not chunk:
                    selector.unregister(key.fd)
                    continue
                buffer = buffers[key.fd]
                buffer.extend(chunk)
                if len(buffer) > _INSTALL_OUTPUT_BYTES:
                    raise AgentDeliveryError(
                        f"Muse skill installation output exceeds {_INSTALL_OUTPUT_BYTES} bytes"
                    )
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise subprocess.TimeoutExpired(command, _INSTALL_TIMEOUT)
        returncode = process.wait(timeout=remaining)
    except BaseException:
        try:
            os.killpg(process.pid, signal.SIGKILL)
        except ProcessLookupError:
            pass
        try:
            process.kill()
        except ProcessLookupError:
            pass
        process.wait()
        raise
    finally:
        selector.close()
    try:
        stdout = bytes(buffers[process.stdout.fileno()]).decode("utf-8", errors="strict")
        stderr = bytes(buffers[process.stderr.fileno()]).decode("utf-8", errors="strict")
    except UnicodeError as exc:
        raise AgentDeliveryError(f"Muse skill installation returned invalid UTF-8: {exc}") from exc
    return returncode, stdout, stderr


def _install_muse(content: str, *, force: bool) -> str:
    # Validate code before making any configuration directories.
    executable = _muse_executable()
    config_home = _muse_config_home()
    skills = config_home / "skills"
    _ensure_real_directory(skills)
    destination = skills / "agentctl" / "SKILL.md"
    _refuse_directory_symlink(destination.parent)
    outcome = _existing(destination, content, force=force)
    if outcome is not None:
        return outcome
    with tempfile.TemporaryDirectory(prefix="agentctl-muse-skill-") as temporary:
        source = Path(temporary) / "agentctl"
        source.mkdir(mode=0o700)
        source_file = source / "SKILL.md"
        source_file.write_text(content, encoding="utf-8")
        source_file.chmod(0o600)
        command = [
            executable, "skills", "install", str(source), "--scope", "user",
            "--name", "agentctl", "--json",
        ]
        if force:
            command.append("--force")
        environment = dict(os.environ)
        environment["XDG_CONFIG_HOME"] = str(config_home.parent)
        try:
            returncode, stdout, stderr = _run_muse_installer(command, environment)
        except (OSError, UnicodeError, subprocess.TimeoutExpired) as exc:
            raise AgentDeliveryError(f"Muse skill installation failed: {exc}") from exc
    if returncode != 0:
        detail = (stderr or stdout).strip()[-4000:]
        raise AgentDeliveryError(f"Muse skill installation failed: {detail or returncode}")
    try:
        receipt = json.loads(stdout)
    except (json.JSONDecodeError, RecursionError) as exc:
        raise AgentDeliveryError(f"Muse skill installation returned invalid JSON: {exc}") from exc
    if (not isinstance(receipt, dict) or not isinstance(receipt.get("installed"), dict)
            or receipt["installed"].get("id") != "agentctl"):
        raise AgentDeliveryError("Muse skill installation returned no agentctl receipt")
    if _existing(destination, content, force=False) != "unchanged":
        raise AgentDeliveryError("Muse reported installation without the expected managed skill")
    return "installed"


def install_skill(harnesses: Sequence[str], *, force: bool = False) -> dict[str, list[str]]:
    """Install one byte-identical bundled skill for each selected harness."""
    selected = tuple(dict.fromkeys(harnesses or HARNESSES))
    invalid = sorted(set(selected) - set(HARNESSES))
    if invalid:
        raise AgentDeliveryError(f"unsupported skill harnesses: {', '.join(invalid)}")
    content = files("agentctl").joinpath("AGENTCTL_SKILL.md").read_text(encoding="utf-8")
    result: dict[str, list[str]] = {"installed": [], "unchanged": []}
    for harness in selected:
        if harness == "muse":
            outcome = _install_muse(content, force=force)
        else:
            destination = _root(harness) / "agentctl" / "SKILL.md"
            outcome = _write(destination, content, force=force)
        result[outcome].append(harness)
    return result

File: rs/agentctl/src/skill_install.rs
//! Idempotent installation of the bundled agentctl skill.
use serde::Serialize;
use serde_json::Value;
use std::fs::{self, DirBuilder, OpenOptions};
use std::io::Write as _;
use std::os::unix::fs::{DirBuilderExt, OpenOptionsExt, PermissionsExt};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use crate::agent::AgentError;

fn fail(message: impl Into<String>) -> AgentError {
    AgentError::Delivery(message.into())
}

#[derive(Debug, Serialize)]
pub(crate) struct InstallResult {
    installed: Vec<String>,
    unchanged: Vec<String>,
}

fn direct_root(harness: &str) -> Result<PathBuf, AgentError> {
    if let Some(value) = std::env::var_os(format!(
        "AGENTCTL_{}_SKILLS_DIR",
        harness.to_ascii_uppercase()
    )) {
        let root = PathBuf::from(value);
        if !root.is_absolute() {
            return Err(fail(format!(
                "{harness} skill directory override must be absolute"
            )));
        }
        return Ok(root);
    }
    let directory = match harness {
        "codex" => ".codex",
        "claude" => ".claude",
        _ => {
            return Err(fail(format!(
                "unsupported direct skill harness {harness:?}"
            )))
        }
    };
    Ok(crate::client::account_home()?
        .join(directory)
        .join("skills"))
}

fn ensure_real_directory(path: &Path) -> Result<(), AgentError> {
    if !path.is_absolute() {
        return Err(fail(format!(
            "skill destination must be absolute: {}",
            path.display()
        )));
    }
    let mut current = PathBuf::new();
    for component in path.components() {
        current.push(component.as_os_str());
        match fs::symlink_metadata(&current) {
            Ok(metadata) if metadata.file_type().is_symlink() || !metadata.is_dir() => {
                return Err(fail(format!(
                    "refusing non-directory skill destination: {}",
                    current.display()
                )));
            }
            Ok(_) => {}
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
                DirBuilder::new()
                    .mode(0o700)
                    .create(&current)
                    .map_err(|error| {
                        fail(format!(
                            "cannot create skill destination {}: {error}",
                            current.display()
                        ))
                    })?;
            }
            Err(error) => {
                return Err(fail(format!(
                    "cannot inspect skill destination {}: {error}",
                    current.display()
                )));
            }
        }
    }
    Ok(())
}

fn existing(path: &Path, force: bool) -> Result<Option<&'static str>, AgentError> {
    let metadata = match fs::symlink_metadata(path) {
        Ok(metadata) => metadata,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
        Err(error) => {
            return Err(fail(format!(
                "cannot inspect installed skill {}: {error}",
                path.display()
            )));
        }
    };
    if metadata.file_type().is_symlink() || !metadata.is_file() {
        return Err(fail(format!(
            "refusing non-regular skill file: {}",
            path.display()
        )));
    }
    let bytes = fs::read(path).map_err(|error| {
        fail(format!(
            "cannot inspect installed skill {}: {error}",
            path.display()
        ))
    })?;
    if bytes == crate::AGENTCTL_SKILL.as_bytes() {
        return Ok(Some("unchanged"));
    }
    if !force {
        return Err(fail(format!(
            "refusing to overwrite divergent skill {}; inspect it or pass --force",
            path.display()
        )));
    }
    Ok(None)
}

fn write_direct(path: &Path, force: bool) -> Result<&'static str, AgentError> {
    let directory = path.parent().expect("skill parent");
    ensure_real_directory(directory)?;
    if let Some(outcome) = existing(path, force)? {
        return Ok(outcome);
    }
    let nonce = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|error| fail(error.to_string()))?
        .as_nanos();
    let temporary = directory.join(format!(".SKILL.md.{}.{nonce}", std::process::id()));
    let mut file = OpenOptions::new()
        .create_new(true)
        .write(true)
        .mode(0o600)
        .open(&temporary)
        .map_err(|error| fail(error.to_string()))?;
    let result = (|| {
        file.write_all(crate::AGENTCTL_SKILL.as_bytes())?;
        file.sync_all()?;
        fs::rename(&temporary, path)?;
        fs::File::open(directory)?.sync_all()?;
        Ok::<(), std::io::Error>(())
    })();
    if let Err(error) = result {
        let _ = fs::remove_file(&temporary);
        return Err(fail(error.to_string()));
    }
    Ok("installed")
}

fn muse_config_home() -> Result<PathBuf, AgentError> {
    if let Some(value) = std::env::var_os("XDG_CONFIG_HOME") {
        let root = PathBuf::from(value);
        if !root.is_absolute() {
            return Err(fail(
                "XDG_CONFIG_HOME must be absolute for Muse skill installation",
            ));
        }
        return Ok(root.join("muse"));
    }
    Ok(crate::client::account_home()?.join(".config/muse"))
}

fn muse_executable() -> Result<PathBuf, AgentError> {
    let Some(configured) = std::env::var_os("AGENTCTL_MUSE_BIN") else {
        return Ok(crate::client::resolve_harness_executable("muse")?);
    };
    let path = PathBuf::from(configured);
    if !path.is_absolute() {
        return Err(fail("AGENTCTL_MUSE_BIN must be an absolute path"));
    }
    let resolved = fs::canonicalize(&path).map_err(|error| {
        fail(format!(
            "cannot inspect Muse executable {}: {error}",
            path.display()
        ))
    })?;
    let metadata = fs::metadata(&resolved).map_err(|error| {
        fail(format!(
            "cannot inspect Muse executable {}: {error}",
            resolved.display()
        ))
    })?;
    if !metadata.is_file()
        || metadata.permissions().mode() & 0o111 == 0
        || metadata.permissions().mode() & 0o022 != 0
    {
        return Err(fail(format!(
            "refusing unsafe Muse executable: {}",
            resolved.display()
        )));
    }
    Ok(resolved)
}

fn output_detail(output: &[u8]) -> String {
    let start = output.len().saturating_sub(4000);
    String::from_utf8_lossy(&output[start..]).trim().to_owned()
}

fn install_muse(force: bool) -> Result<&'static str, AgentError> {
    // Resolve and validate code before making any configuration directories.
    let executable = muse_executable()?;
    let config_home = muse_config_home()?;
    let skills = config_home.join("skills");
    ensure_real_directory(&skills)?;
    let destination_directory = skills.join("agentctl");
    if let Ok(metadata) = fs::symlink_metadata(&destination_directory) {
        if metadata.file_type().is_symlink() || !metadata.is_dir() {
            return Err(fail(format!(
                "refusing non-directory skill destination: {}",
                destination_directory.display()
            )));
        }
    }
    let destination = destination_directory.join("SKILL.md");
    if let Some(outcome) = existing(&destination, force)? {
        return Ok(outcome);
    }

    let nonce = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|error| fail(error.to_string()))?
        .as_nanos();
    let staging = std::env::temp_dir().join(format!(
        "agentctl-muse-skill.{}.{nonce}",
        std::process::id()
    ));
    DirBuilder::new()
        .mode(0o700)
        .create(&staging)
        .map_err(|error| fail(format!("cannot stage Muse skill: {error}")))?;
    let source = staging.join("agentctl");
    DirBuilder::new()
        .mode(0o700)
        .create(&source)
        .map_err(|error| fail(format!("cannot stage Muse skill: {error}")))?;
    let source_file = source.join("SKILL.md");
    let write_result = (|| {
        let mut file = OpenOptions::new()
            .create_new(true)
            .write(true)
            .mode(0o600)
            .open(&source_file)?;
        file.write_all(crate::AGENTCTL_SKILL.as_bytes())?;
        file.sync_all()?;
        Ok::<(), std::io::Error>(())
    })();
    if let Err(error) = write_result {
        let _ = fs::remove_dir_all(&staging);
        return Err(fail(format!("cannot stage Muse skill: {error}")));
    }

    let mut command = Command::new(executable);
    command
        .args(["skills", "install"])
        .arg(&source)
        .args(["--scope", "user", "--name", "agentctl", "--json"])
        .env(
            "XDG_CONFIG_HOME",
            config_home.parent().expect("Muse config parent"),
        )
        .stdin(Stdio::null());
    if force {
        command.arg("--force");
    }
    let output = crate::client::bounded_output(command, Duration::from_secs(30));
    let _ = fs::remove_dir_all(&staging);
    let output =
        output.map_err(|error| fail(format!("Muse skill installation failed: {error}")))?;
    if !output.status.success() {
        let detail = if output.stderr.is_empty() {
            output_detail(&output.stdout)
        } else {
            output_detail(&output.stderr)
        };
        return Err(fail(format!(
            "Muse skill installation failed: {}",
            if detail.is_empty() {
                output.status.to_string()
            } else {
                detail
            }
        )));
    }
    let receipt: Value = serde_json::from_slice(&output.stdout).map_err(|error| {
        fail(format!(
            "Muse skill installation returned invalid JSON: {error}"
        ))
    })?;
    if receipt
        .get("installed")
        .and_then(Value::as_object)
        .and_then(|value| value.get("id"))
        .and_then(Value::as_str)
        != Some("agentctl")
    {
        return Err(fail("Muse skill installation returned no agentctl receipt"));
    }
    if existing(&destination, false)? != Some("unchanged") {
        return Err(fail(
            "Muse reported installation without the expected managed skill",
        ));
    }
    Ok("installed")
}

pub(crate) fn install(harnesses: &[String], force: bool) -> Result<InstallResult, AgentError> {
    let selected = if harnesses.is_empty() {
        vec!["codex".to_owned(), "claude".to_owned(), "muse".to_owned()]
    } else {
        let mut values = Vec::new();
        for harness in harnesses {
            if !matches!(harness.as_str(), "codex" | "claude" | "muse") {
                return Err(fail(format!("unsupported skill harness {harness:?}")));
            }
            if !values.contains(harness) {
                values.push(harness.clone());
            }
        }
        values
    };
    let mut result = InstallResult {
        installed: Vec::new(),
        unchanged: Vec::new(),
    };
    for harness in selected {
        let outcome = if harness == "muse" {
            install_muse(force)?
        } else {
            let destination = direct_root(&harness)?.join("agentctl/SKILL.md");
            write_direct(&destination, force)?
        };
        match outcome {
            "installed" => result.installed.push(harness),
            "unchanged" => result.unchanged.push(harness),
            _ => unreachable!(),
        }
    }
    Ok(result)
}
